File: leet-code/15. 3sum/ThreeSum15.java
import java.util.Arrays;
import java.util.HashSet;
import java.util.List;
import java.util.Set;

public class ThreeSum15 {
    public List<List<Integer>> threeSum(int[] nums) {
        Arrays.sort(nums);
        Set<List<Integer>> answer = new HashSet<>();

        for (int i = 0; i < nums.length - 2; i++) {
            int left = i + 1;
            int right = nums.length - 1;

            while (left < right) {
                int sums = nums[i] + nums[left] + nums[right];

                if (sums < 0) {
                    left++;
                } else if (sums > 0) {
                    right--;
                } else {
                    answer.add(List.of(nums[i], nums[left], nums[right]));
                    left++;
                    right--;
                }
            }
        }

        System.out.println(answer.stream().toList());
        return answer
                .stream()
                .toList();
    }
}

File: leet-code/15. 3sum/ThreeSum15.kt
package com.example.algorithm

import kotlin.collections.ArrayList

class ThreeSum15 {
    fun threeSum(nums: IntArray): List<List<Int>> {
        nums.sort()
        val answer: MutableList<List<Int>> = ArrayList()
        for (i in 0 until nums.size - 2) {
            if (i > 0 && nums[i] == nums[i - 1]) continue
            var left = i + 1
            var right = nums.size - 1
            while (left < right) {
                val sum = nums[i] + nums[left] + nums[right]
                if (sum > 0) {
                    right -= 1
                } else if (sum < 0) {
                    left += 1
                } else {
                    answer.add(listOf(nums[i], nums[left], nums[right]))
                    while (left < right && nums[left] == nums[left + 1]) left += 1
                    while (left < right && nums[right] == nums[right - 1]) right -= 1
                    left += 1
                    right -= 1
                }
            }
        }
        return answer
    }
}